bitswap/server/internal/decision/engine.go

```diff
@@ -827,19 +798,91 @@ func (e *Engine) MessageReceived(ctx context.Context, p peer.ID, m bsmsg.BitSwap
 		})
 	}
 
-	// Push entries onto the request queue
-	if len(activeEntries) > 0 {
+	// Push entries onto the request queue and signal network that new work is ready.
+	if len(activeEntries) != 0 {
 		e.peerRequestQueue.PushTasksTruncated(e.maxQueuedWantlistEntriesPerPeer, p, activeEntries...)
 		e.updateMetrics()
+		e.signalNewWork()
 	}
 	return false
 }
 
+// handleOverflow processes incoming wants that could not be addded to the peer
+// ledger without exceeding the peer want limit. These are handled by trying to
+// make room by canceling existing wants for which there is no block. If this
+// does not make sufficient room, then any lower priority wants that have
+// blocks are canceled.
+func (e *Engine) handleOverflow(ctx context.Context, p peer.ID, wants, overflow []bsmsg.Entry) []bsmsg.Entry {
+	existingWants := e.peerLedger.WantlistForPeer(p)
+	// Sort wl and overflow from least to most important.
+	slices.SortFunc(existingWants, func(a, b wl.Entry) int {
+		return cmp.Compare(a.Priority, b.Priority)
+	})
+	slices.SortFunc(overflow, func(a, b bsmsg.Entry) int {
+		return cmp.Compare(a.Entry.Priority, b.Entry.Priority)
+	})
+
+	queuedWantKs := cid.NewSet()
+	for _, entry := range existingWants {
+		queuedWantKs.Add(entry.Cid)
+	}
+	queuedBlockSizes, err := e.bsm.getBlockSizes(ctx, queuedWantKs.Keys())
+	if err != nil {
+		log.Info("aborting overflow processing", err)
+		return wants
+	}
+
+	// Remove entries for blocks that are not present to make room for overflow.
+	var removed []int
+	for i, w := range existingWants {
+		if _, found := queuedBlockSizes[w.Cid]; !found {
+			// Cancel lowest priority dont-have.
+			if e.peerLedger.CancelWant(p, w.Cid) {
+				e.peerRequestQueue.Remove(w.Cid, p)
+			}
+			removed = append(removed, i)
+			// Add highest priority overflow.
+			lastOver := overflow[len(overflow)-1]
+			overflow = overflow[:len(overflow)-1]
+			e.peerLedger.Wants(p, lastOver.Entry, 0)
+			wants = append(wants, lastOver)
+			if len(overflow) == 0 {
+				break
+			}
+		}
+	}
+
+	// Not enough dont-haves removed. Replace existing entries, that are a
+	// lower priority, with overflow entries.
```

Comment:
It probably will be unlikely to happen given that the client decreases priority over time. The thinking here was looking more toward a future where priority would be set by path distance, where items closer to a DAG root have a higher priority. In that case it seems likely that as new wants are requested that some do have a high priority because they are root or closer to root items.

1, There are two short circuits happening here. To illustrate, lets start with no short circuits:
```
for _, o := range overflow {
    for _, w := range wantlist {
       if o.Priority > w.Priority {
           replace(w, 0) // replace w with o
       }
   }
}
```
The first short is to get rid of the inner loop. That can be done by only looking at the first item in the wantlist, since the first item has the lowest remainnig priority and therefore is most replaceable. That gives:
```
for _, o := range overflow {
    for _, w := range wantlist {
       if o.Priority > w.Priority {
           replace(w, 0) // replace w with o
       }
   }
}
```